docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,125 @@
+
+
+## Motivation
+
+Deployment hooks are needed to provide users with a way to execute arbitrary commands necessary to complete a deployment, targeting executables which may be maintained outside the docker image being deployed.
+
+Goals of this design:
+
+1. Identify deployment hook use cases
+2. Define the integration of deployment hooks and the deployment lifecycle
+3. Describe a deployment hook API
+
+
+## Comparison of potential approaches
+
+##### Upstream container lifecycle hooks
+
+Kubernetes provides *container lifecycle hooks* for containers within a Pod. Currently, post-start and pre-stop hooks are supported. For deployments, post-start is the most relevant. Because these hooks are implemented by the Kubelet, the post-start hook provides some unique guarantees:
+
+1. The hook is executed synchronously during the pod lifecycle.
+2. The status of the pod is linked to the status and outcome of the hook execution.
+  1. The pod will not enter a ready status until the hook has completed successfully.
+    1. Service endpoints will not be created for the pod until the pod has a ready status.
+  1. If the hook fails, the pod's creation is considered a failure, and the retry behavior is restart-policy driven in the usual way.
+
+Because deployments are represented as replication controllers, lifecycle hooks defined for containers are executed for every container in the replica set for the deployment. This behavior has complexity implications when applied to deployment use cases:
+
+1. The hooks for all pods in the deployment will race, placing a burden on hook authors (e.g., the hooks would generally need to be tolerant of concurrent hook execution and implement manual coordination.)
+
+
+##### Deployment hooks
+
+Deployment hooks implemented as run-once pods external to the Kubelet provide a different set of guarantees:
+
+1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
+  1. Races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
+2. Hooks defined in terms of deployments are conceptually easier to reason about from the perspective of a user defining a deployment workflow.
+
+When implementing a hook which runs after a deployment has been scaled up, there are special considerations imposed by such an external hook mechanism:
+
+1. Nothing prevents external connectivity races: when a the deployment's pods are ready, they become routable to services and other pods within an application the moment their containers enter a ready state, likely before or during or before hook execution.
+2. Hook execution can't be atomically linked to the deployment pods' statuses: If a hook failure should result in deployment failure, a previously scaled and exposed application must be rolled back when ideally the application shouldn't have been exposed prior to hook success.
```

Comment:
>1) run my hook before the deployment (and potentially abort the deployment if the hook fails)

Basically - but a subtle clarification: the deployment would exist (with a replica count 0) but wouldn't be scaled (the strategy wouldn't run) until the hook completed.

>2) run my hook exactly once, after the deployment (challenges with the hook not completing before the deployment is exposed, or the hook attempting to run before the deployment is truly up)

Yeah, "after the deployment" means the strategy could have scaled up the deployment, and now it could be either partially or fully deployed and exposed at any time.

>3) run my hook on every pod create event (implicitly including scale up events) (the existing k8s post start functionality)

Yes.

>and the issues listed here would apply to both item 2+ item 3?

The issues listed in this specific part apply to approach 2. Approach 3 has its own issues detailed in the section above dedicated to 3.